File: leetcode/7-number-of-submatrices-that-sum-to-target/myHan.java
class Solution {
    public int numSubmatrixSumTarget(int[][] matrix, int target) {
        int ans = 0;
        int[][] acc = new int[matrix.length+1][matrix[0].length+1];
        for(int r = 0; r < matrix.length; r++) {
            for(int c = 0; c < matrix[0].length; c++) {
                acc[r+1][c+1] = matrix[r][c] + acc[r+1][c] + acc[r][c+1] - acc[r][c];
            }
        }
        for(int r1 = 0; r1 < matrix.length; r1++) {
            for(int c1 = 0; c1 < matrix[0].length; c1++) {
                for(int r2 = r1+1; r2 <= matrix.length; r2++) {
                    for(int c2 = c1+1; c2 <= matrix[0].length; c2++) {
                        if(acc[r2][c2] - acc[r2][c1] - acc[r1][c2] + acc[r1][c1] == target) {
                            ans++;
                        }
                    }
                }
            }
        }
        return ans;
    }
}

File: leetcode/20-redundant-connection/myHan.ex
defmodule Solution do
	@doc """
	Find a representative of the group

	## Examples

	    iex> Solution.find(%{1 => 2, 2 => 3, 3 => 3}, 1)
	    {%{1 => 3, 2 => 3, 3 => 3}, 3}
	"""
	def find(disjoint_set, u) do
		v = Map.get(disjoint_set, u)
		cond do
			u == v -> {disjoint_set, u}
			true   ->
				{disjoint_set, g} = find(disjoint_set, v)
				{Map.put(disjoint_set, u, g), g}
		end
	end

	def union(disjoint_set, u, v) do
		Map.put(disjoint_set, u, v)
	end

	@spec find_redundant_connection(edges :: [[integer]]) :: [integer]
	def find_redundant_connection(edges) do
		disjoint_set = 1..length(edges) |> Enum.reduce(%{}, fn u, acc -> Map.put(acc, u, u) end)

    	{_, ans} = Enum.reduce(edges, {disjoint_set, nil}, fn
    		[u, v], {disjoint_set, nil} ->
    			{disjoint_set, gu} = find(disjoint_set, u)
    			{disjoint_set, gv} = find(disjoint_set, v)
    			cond do
    				gu == gv -> {disjoint_set, [u, v]}
    				true -> {union(disjoint_set, gu, gv), nil}
    			end
    		_, ret          -> ret
    	end)
    	ans
    end
end
